std/machines/hash/poseidon2_bb.asm

```diff
@@ -0,0 +1,288 @@
+    link if is_used ~> (input_high[0], input_low[0]) = mem.mload(input_addr_high[0], input_addr_low[0], time_step);
+    link if is_used ~> (input_high[1], input_low[1]) = mem.mload(input_addr_high[1], input_addr_low[1], time_step);
+    link if is_used ~> (input_high[2], input_low[2]) = mem.mload(input_addr_high[2], input_addr_low[2], time_step);
+    link if is_used ~> (input_high[3], input_low[3]) = mem.mload(input_addr_high[3], input_addr_low[3], time_step);
+    link if is_used ~> (input_high[4], input_low[4]) = mem.mload(input_addr_high[4], input_addr_low[4], time_step);
+    link if is_used ~> (input_high[5], input_low[5]) = mem.mload(input_addr_high[5], input_addr_low[5], time_step);
+    link if is_used ~> (input_high[6], input_low[6]) = mem.mload(input_addr_high[6], input_addr_low[6], time_step);
+    link if is_used ~> (input_high[7], input_low[7]) = mem.mload(input_addr_high[7], input_addr_low[7], time_step);
+    link if is_used ~> (input_high[8], input_low[8]) = mem.mload(input_addr_high[8], input_addr_low[8], time_step);
+    link if is_used ~> (input_high[9], input_low[9]) = mem.mload(input_addr_high[9], input_addr_low[9], time_step);
+    link if is_used ~> (input_high[10], input_low[10]) = mem.mload(input_addr_high[10], input_addr_low[10], time_step);
+    link if is_used ~> (input_high[11], input_low[11]) = mem.mload(input_addr_high[11], input_addr_low[11], time_step);
+    link if is_used ~> (input_high[12], input_low[12]) = mem.mload(input_addr_high[12], input_addr_low[12], time_step);
+    link if is_used ~> (input_high[13], input_low[13]) = mem.mload(input_addr_high[13], input_addr_low[13], time_step);
+    link if is_used ~> (input_high[14], input_low[14]) = mem.mload(input_addr_high[14], input_addr_low[14], time_step);
+    link if is_used ~> (input_high[15], input_low[15]) = mem.mload(input_addr_high[15], input_addr_low[15], time_step);
+
+    let input = array::zip(input_low, input_high, |low, high| low + 0x10000 * high);
+
+    // Perform the inital MDS step
+    let pre_rounds = apply_mds(input, STATE_SIZE);
+
+    // Perform most of the rounds
+    let final_full_state = (constr || {
+        // Perform the first half of the external rounds
+        let after_initial_rounds = utils::fold(
+            HALF_EXTERNAL_ROUNDS, |round_idx| round_idx, pre_rounds,
+            constr |pre_state, round_idx| {
+            //    let post_state: col[STATE_SIZE];
+                let post_state = array::new(STATE_SIZE, |_| { let x; x});
+                external_round(round_idx, pre_state, post_state);
+                post_state
+            }
+        );
+
+        // Perform the internal rounds
+        let after_internal_rounds = utils::fold(
+            INTERNAL_ROUNDS, |round_idx| round_idx, after_initial_rounds,
+            constr |pre_state, round_idx| {
+                let post_state = array::new(STATE_SIZE, |_| { let x; x});
```

Comment:
Again, we should optimize this later, but do you think we have to commit to all the intermediate states? Could `{internal,external}_round()` return an expression and we just use that?

Since Poseidon is an algebraic hash function, I think we could in principle express the entire function as one big algebraic expression. The only problem would be that its would be of too high degree. But eventually we should try to minimize the witness columns.

Reply:
Yes.

See the issue I opened related to this: #1971

Reply:
Actually, not really. If you multiply 2 polynomials, the resulting polynomial has size m*n, so they can grow to intractable sizes.

Unless we internally represent polynomials as a DAG of multiplication of polynomial factors (which I guess neither we nor the backends do).

Reply:
> If you multiply 2 polynomials, the resulting polynomial has size m*n, so they can grow to intractable sizes.

I don't understand this comment! If you multiply two polynomials of degrees m and n, you get a polynomial of degree m + n, right?

I do think that #1971 would fix this: We'd add each intermediate state as an intermediate polynomials, which depending on the degree bound would be inlined or materialized as a witness column.

Reply:
Oh, maybe you're talking about the fact that if e.g. if you repeatedly compute an S-Box $x^7$ and $x$ is an expression and you inline everything, the resulting expression grows exponentially? That's fixed by intermediate columns and Thibaut just pushed #1995, which makes the Plonky3 backend aware of them :)